### PR TITLE
this is template cleanup and jenkins pipeline

### PR DESCRIPTION
**Pull request by goern**
... :)

### CODE REVIEW

File: openshift/conf.sh [hunk 1]
```diff
@@ -11,6 +11,8 @@ export THOTH_LOCAL=1
 
 # Adjust if you deploy to actual cluster - this is a namespace to separate different deployments:
 export THOTH_DEPLOYMENT_NAME="local-$USER"
+export THOTH_MIDDLEEND_NAMESPACE="thoth-test-dev"
+export THOTH_BACKEND_NAMESPACE="thoth-test-dev"
```
Comment (fridex) [inline openshift/conf.sh]: If we want to move this to conf.sh, maybe it will be a good idea to place all namespace configuration here - there is missing THOHT_FRONTED_NAMESPACE.

File: openshift/deploy.sh [hunk 2]
```diff
@@ -16,7 +16,7 @@ fi
 
 [ \( "${THOTH_LOCAL}" \) ] && oc cluster up --routing-suffix="${THOTH_ROUTING_SUFFIX}"
 oc new-project thoth-frontend
-oc new-project thoth-middleend
+oc new-project ${THOTH_MIDDLEEND_NAMESPACE}
 oc new-project thoth-backend
```
Comment (fridex) [inline openshift/deploy.sh]: The conf.sh file is sourced above. As there are configured THOTH_*_NAMESPACE parameters, I would suggest to use them here - if the given namespace already exists, the script will simply print error message and continue with execution.

File: openshift/deploy.sh [hunk 3]
```diff
@@ -25,7 +25,11 @@ oc process -f template.yaml \
         -p THOTH_CEPH_SECRET_KEY=${THOTH_CEPH_SECRET_KEY} \
         -p THOTH_CEPH_HOST=${THOTH_CEPH_HOST} \
         -p THOTH_SECRET=${THOTH_SECRET} \
+		-p THOTH_MIDDLEEND_NAMESPACE=${THOTH_MIDDLEEND_NAMESPACE} \
+		-p THOTH_BACKEND_NAMESPACE=${THOTH_BACKEND_NAMESPACE} \
```
Comment (fridex) [inline openshift/deploy.sh]: Missing THOTH_FRONTEND_NAMESPACE, also indentation is broken.

File: openshift/template.yaml [hunk 4]
```diff
@@ -36,6 +36,7 @@ objects:
   data:
     middletier_namespace: ${THOTH_MIDDLEEND_NAMESPACE}
     backend_namespace: ${THOTH_BACKEND_NAMESPACE}
+    storage-bucket-name: ${THOTH_DEPLOYMENT_NAME}
```
Comment (fridex) [inline openshift/template.yaml]: Personally, I would keep this as deployment-name, the fact that it is also bucket name is transparent now - we will probably use different bucket naming convention in the near future once Sherard comes up with some bucket and prefix naming schema. Here and in the parameter definition below.

File: openshift/template.yaml [hunk 5]
```diff
@@ -5,304 +5,216 @@ labels:
...
 #
-# Secrets
-#
+# Application Configuration
+## Secrets
 - apiVersion: v1
   kind: Secret
   metadata:
-    name: secret
-    namespace: ${THOTH_FRONTEND_NAMESPACE}
```
Comment (fridex) [inline openshift/template.yaml]: Why was namespace deleted? Will these changes made for the transition to configmaps work when we will deploy thoth to multiple namespaces as it is designed?

File: openshift/template.yaml [hunk 6]
```diff
@@ -5,304 +5,216 @@ labels:
...
 #
 # Services
-#
 - apiVersion: v1
   kind: Service
   metadata:
     labels:
-      service: thoth-user-api
-    name: thoth-user-api
-    namespace: ${THOTH_FRONTEND_NAMESPACE}
```
Comment (fridex) [inline openshift/template.yaml]: Did you try to deploy thoth to multiple namespaces? I think this will not work. Here and other parts below.

File: openshift/template.yaml [hunk 7]
```diff
@@ -5,304 +5,216 @@ labels:
...
         - env:
           - name: APP_SERVICE_PORT
-            value: "8080"
+            valueFrom:
+              configMapKeyRef:
+                key: user-api-service-port
+                name: thoth
           - name: KUBERNETES_API_URL
-            value: ${KUBERNETES_API_URL}
+            value: 'https://kubernetes.default.svc.cluster.local'
```
Comment (fridex) [inline openshift/template.yaml]: I would like to keep this configurable. If we will get some kubernetes cluster (e.g. hosted on GKE) that we would like to orchestrate from user API service, we can do so by adjusting configuration without a need to deploy whole thoth-core. It will be just a pool of resources we could use for middleend/backaned. The same applies to result API.

File: openshift/template.yaml [hunk 8]
```diff
@@ -5,304 +5,216 @@ labels:
...
-            value: ${THOTH_MIDDLEEND_NAMESPACE}
           - name: THOTH_RESULT_API_HOSTNAME
-            value: ${THOTH_RESULT_API_HOSTNAME}
+            value: thoth-result-api
           - name: THOTH_RESULT_API_URL
-            value: http://thoth-result-api-hostname-${THOTH_MIDDLEEND_NAMESPACE}.${THOTH_ROUTING_SUFFIX}
+            value: http://thoth-result-api
           - name: THOTH_MIDDLEEND_POD_CPU_LIMIT
-            value: ${THOTH_MIDDLEEND_POD_CPU_LIMIT}
+            value: '0.5'
```
Comment (fridex) [inline openshift/template.yaml]: Any particular reason why this is not in a config map?

File: openshift/template.yaml [hunk 9]
```diff
@@ -414,126 +332,82 @@ objects:
...
-          kind: ImageStreamTag
-          name: 'thoth-result-api:latest'
-
-- apiVersion: v1
-  kind: DeploymentConfig
-  metadata:
-    labels:
-      service: graphexp
-      app: graphexp
-    name: graphexp
```
Comment (fridex) [inline openshift/template.yaml]: Why do we want to remove graphexp?

File: openshift/template.yaml [hunk 10]
```diff
@@ -952,24 +639,12 @@ parameters:
...
-
 - description: Secret to interact with Thoth endpoints that require higher privileges.
   displayName: Thoth secret
   required: true
   name: THOTH_SECRET
 
-- description: Routing suffix of deployment.
-  displayName: Routing suffix
-  required: true
-  name: THOTH_ROUTING_SUFFIX
```
Comment (fridex) [inline openshift/template.yaml]: Why was this removed?

File: openshift/template.yaml [hunk 11]
```diff
@@ -414,126 +332,82 @@ objects:
...
+    name: cleanup-job
   spec:
     #schedule: "@weekly"
     schedule: "*/1 * * * *"
     jobTemplate:
       spec:
         template:
           spec:
-            serviceAccountName: master-access
+            serviceAccountName: analyzer
```
Comment (fridex) [inline openshift/template.yaml]: The cleanup job is not analyzer so maybe this is confusing.

**Comment by fridex** [comment]
I've tried to deploy with these changes using the deploy.sh script and the deployment failed: 
```
+ oc process -f template.yaml -p THOTH_ROUTING_SUFFIX=172.17.0.1.nip.io -p THOTH_DEPLOYMENT_NAME=local-fpokorny -p THOTH_CEPH_KEY_ID= -p THOTH_CEPH_SECRET_KEY= -p THOTH_CEPH_HOST= -p THOTH_SECRET=secret -p THOTH_MIDDLEEND_NAMESPACE=thoth-test-dev -p THOTH_BACKEND_NAMESPACE=thoth-test-dev
error: unknown parameter name "THOTH_ROUTING_SUFFIX"
error: no objects passed to apply
```

I would like to keep a single script deployment to be still usable even for `oc cluster up` deployment so we can easily deploy thoth wherever/whenever we need and deployment will be reproducible anytime (not like in other projects in which I used to be  :) ).

File: openshift/template.yaml (hunk 5 above)
Comment (goern) [reply openshift/template.yaml]: thats right, secrets are only accessible from within the same namespace, I gonna fix it!

File: openshift/template.yaml (hunk 7 above)
Comment (goern) [reply openshift/template.yaml]: But the cluster internal API URL of Kubernetes is generic, why should we configure it?

File: openshift/template.yaml (hunk 9 above)
Comment (goern) [reply openshift/template.yaml]: GraphExp should be part of a website, not anything we deploy with the core service

File: openshift/template.yaml (hunk 11 above)
Comment (goern) [reply openshift/template.yaml]: right, who uses it for what?

File: openshift/template.yaml (hunk 7 above)
Comment (fridex) [reply openshift/template.yaml]: If we will have a hosted cluster somewhere with some specific hardware, we don't need to deploy the whole thoth there, but we can configure api service to submit analyzes to this cluster by providing kubernetes API and kubernetes token - the same way how oc binary communicates with openshift.

This is internal cluster url, but it can be also external where we would also supply API token to communicate with kubernetes master [1] (defaults to service account).

[1] https://github.com/fridex/thoth-user-api/blob/master/thoth_user_api/configuration.py#L24

File: openshift/template.yaml (hunk 11 above)
Comment (fridex) [reply openshift/template.yaml]: It is used by the user API service to run pods on analysis request as well as by the cleanup-job that cleans old analyzes [1].

https://github.com/fridex/thoth-core#the-overall-architecture

File: openshift/template.yaml (hunk 9 above)
Comment (fridex) [reply openshift/template.yaml]: In that case, I would move it to a separate YAML file so if we would like to deploy it, we still have this s2i configuration.

File: openshift/template.yaml (hunk 6 above)
Comment (goern) [reply openshift/template.yaml]: Might by, my focus is to get the Jenkins Pipeline up and running, as that is one namespace anyways... I didnt test it. I have created https://trello.com/c/6V0JAy6L

File: openshift/template.yaml (hunk 7 above)
Comment (goern) [reply openshift/template.yaml]: hmm, understood. This feature doesnt seem to be on the critical path, I would like to focus on getting a CI/CD pipeline up and running

File: openshift/template.yaml (hunk 11 above)
Comment (goern) [reply openshift/template.yaml]: both use cases seem to be related to an analyzer, that is why I called the SA "analyzer" :P

File: openshift/conf.sh (hunk 1 above)
Comment (goern) [reply openshift/conf.sh]: Check out the ansible playbooks, their functionality should be equivalent to conf/deploy and they work with and OpenShift Cluster, just provide OCP_TOKEN and OCP_URL

File: openshift/template.yaml (hunk 7 above)
Comment (fridex) [reply openshift/template.yaml]: This feature is already there - there is only required to adjust parameters of deployment if we would like to do that.

File: openshift/template.yaml (hunk 11 above)
Comment (fridex) [reply openshift/template.yaml]: It's not assigned to analyzer itself, but rather to the other parts that work with analyzers - that's why it sounds confusing to me. Something like analyzer-control or stuff like that would trigger more meaning to "something that handles things with analyzers".

**Comment by goern** [comment]
@fridex pls chk, happy easter!

**Comment by fridex** [comment]
> @fridex pls chk, happy easter!

Happy Easter as well! ;)
